docs: add exclude include check feature tutorial (#667)
oracle/macaron: docs/source/pages/tutorials/exclude_include_checks.rst

.. Copyright (c) 2024 - 2024, Oracle and/or its affiliates. All rights reserved.
.. Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.

=====================================
Exclude and include checks in Macaron
=====================================

.. note::

    This feature is only available from Macaron 0.8.0

This tutorial will show how you can configure Macaron to:

* exclude checks that you don't want to run.
* include only checks that you want to run.

.. note::

    By default, Macaron will run all of its checks.

-------------
Prerequisites
-------------

* You are expected to have gone through :ref:`this tutorial <detect-malicious-java-dep>`.
* This tutorial requires a high-level understanding of checks in Macaron and how they depend on each other. Please see this :ref:`page <macaron-developer-guide>` for more information.

------------------
Motivating example
------------------

> Let's say you want to run the analysis on `the micronaut-projects/micronaut-core Github repository <https://github.com/micronaut-projects/micronaut-core>`_ at the tag ``4.3.10``.

Reviewer: Please use the maven PURL not the repository PURL. Macaron can automatically find the repository and tag.

```suggestion
Let's say you want to run the analysis on Maven artifact `io.micronaut/micronaut-core` version `4.3.10`.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/fab41532fcdc1cf2a52863fea8931998494cacd5

Normally, this is how you would run Macaron:

.. code-block:: shell

>   ./run_macaron.sh analyze --package-url pkg:github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps

Reviewer: ```suggestion
  ./run_macaron.sh analyze --package-url pkg:maven/io.micronaut/micronaut-core@v4.3.10 --skip-deps
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/fab41532fcdc1cf2a52863fea8931998494cacd5

> However, there can be checks in Macaron that are not relevant for ``micronaut-projects/micronaut-core``.

Reviewer: ```suggestion
However, there can be checks in Macaron that are not relevant for the `io.micronaut/micronaut-core` artifact.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/fab41532fcdc1cf2a52863fea8931998494cacd5

> For example, the ``mcn_provenance_witness_level_one_1`` check (defined in :class:`ProvenanceWitnessL1Check <macaron.slsa_analyzer.checks.provenance_witness_l1_check.ProvenanceWitnessL1Check>`) is not relevant because ``micronaut-projects/micronaut-core`` doesn't generate and publish any `Witness provenance <https://github.com/testifysec/witness>`_. Therefore, we could exclude this check from running by performing the following steps.

Reviewer: ```suggestion
For example, the ``mcn_provenance_witness_level_one_1`` check (defined in :class:`ProvenanceWitnessL1Check <macaron.slsa_analyzer.checks.provenance_witness_l1_check.ProvenanceWitnessL1Check>`) is not relevant because ``micronaut-projects/micronaut-core`` generates and publishes :term:`SLSA` provenances and no :term:`Witness` provenances. Therefore, we could exclude this check from running by performing the following steps.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/fab41532fcdc1cf2a52863fea8931998494cacd5

> However, there can be checks in Macaron that are not applicable to ``micronaut-projects/micronaut-core``.
> For example, the ``mcn_provenance_witness_level_one_1`` check (defined in :class:`ProvenanceWitnessL1Check <macaron.slsa_analyzer.checks.provenance_witness_l1_check.ProvenanceWitnessL1Check>`) is not applicable because ``micronaut-projects/micronaut-core`` doesn't generate and publish any `Witness provenance <https://github.com/testifysec/witness>`_. Therefore, we could exclude this check from running by performing the following steps.

Reviewer 2: I would replace "not applicable" with "not relevant", since a possible interpretation of "not applicable" could be that the check cannot be applied, suggesting that not excluding it could cause some issue, which is not the case so a weaker statement of "not relevant" (i.e. you don't care about the result of that check so there's no point running it) seems more appropriate to me.
```suggestion
However, there can be checks in Macaron that are not relevant for ``micronaut-projects/micronaut-core``.
For example, the ``mcn_provenance_witness_level_one_1`` check (defined in :class:`ProvenanceWitnessL1Check <macaron.slsa_analyzer.checks.provenance_witness_l1_check.ProvenanceWitnessL1Check>`) is not relevant because ``micronaut-projects/micronaut-core`` doesn't generate and publish any `Witness provenance <https://github.com/testifysec/witness>`_. Therefore, we could exclude this check from running by performing the following steps.
```

Author: I agree. I will change it to `not relevant`.

1. In your current directory, create a ``defaults.ini`` configuration file with the following content.

.. code-block:: ini

    [analysis.checks]
    exclude = mcn_provenance_witness_level_one_1
    include = *

In the ini configuration above:

* The ``exclude`` option contains the check ID ``mcn_provenance_witness_level_one_1``. The check with this ID will be excluded.

> * The ``include`` option contains the `glob <https://docs.python.org/3/library/glob.html>`_ pattern ``*``. Checks whose ID matched this glob pattern will be included in the analysis. In this case, ``*`` matches all check IDs.

Reviewer 3: ```suggestion
* The ``include`` option contains the `glob <https://docs.python.org/3/library/glob.html>`_ pattern ``*``. Checks with IDs matching this glob pattern will be included in the analysis. In this case, ``*`` matches all check IDs.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/2389ea38eb37dc52a06e7c1b0e15726fb89a6a10

With these two configuration options, all checks except for the excluded ``mcn_provenance_witness_level_one_1`` will run.

2. Run Macaron analyze by providing the path to the custom ``defaults.ini`` configuration.

.. code-block:: shell

>   ./run_macaron.sh --defaults-path ./defaults.ini analyze --package-url pkg:github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps

Reviewer: Please adjust all the commands based on the comment above.

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/fab41532fcdc1cf2a52863fea8931998494cacd5

* If a check is excluded, all of the children which can be transitively reached from it will be excluded.
* If a check is included, all of the parents which can be transitively reached from it will be included.

We define the list of check IDs after extending to transitive parents/children as:

* ``EFFECTIVE_INCLUDED`` = ``EXPLICIT_INCLUDE`` set plus transitive parents
* ``EFFECTIVE_EXCLUDED`` = ``EXPLICIT_EXCLUDE`` set plus transitive children

The final list of checks that will run can be defined as:

* ``FINAL_INCLUDED`` = ``EFFECTIVE_INCLUDED`` minus ``EFFECTIVE_EXCLUDED``

-------------
More examples
-------------

^^^^^^^^^^^^^^^^^^^^^
Not running any check
^^^^^^^^^^^^^^^^^^^^^
The following configurations will result in no check running and Macaron will return on error.

.. code-block:: ini

    [analysis.checks]
    exclude = *
    include = *

.. code-block:: ini

    [analysis.checks]
    exclude =
    include =

^^^^^^^^^^^^^^^^^^^^^^
Run a subset of checks
^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: ini

    [analysis.checks]
    exclude =
    include =
        mcn_provenance_witness_level_one_1
        mcn_trusted_builder_level_three_1

> This will result in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1`` and their transitive parents running.

Reviewer 3: ```suggestion
This will result in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1``, and their transitive parents running.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/2389ea38eb37dc52a06e7c1b0e15726fb89a6a10

^^^^^^^^^^^^^^^^^^^^^^^^^^
Disable a subset of checks
^^^^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: ini

    [analysis.checks]
    exclude =
        mcn_provenance_witness_level_one_1
        mcn_trusted_builder_level_three_1
    include = *

> This will result in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1`` and their transitive children not running.

Reviewer 3: ```suggestion
This will result in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1``, and their transitive children not running.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/2389ea38eb37dc52a06e7c1b0e15726fb89a6a10

* include only checks that you want to run.

------------------------------
Installation and Prerequisites
------------------------------

This tutorial requires a high-level understanding of checks in Macaron and how they depend on each other. Please see this :ref:`page <macaron-developer-guide>` for more information.

Skip this section if you already know how to install Macaron.

.. toggle::

    Please follow the instructions :ref:`here <installation-guide>`. In summary, you need:

        * Docker
        * the ``run_macaron.sh``  script to run the Macaron image.

    .. note:: At the moment, Docker alternatives (e.g. podman) are not supported.

    You also need to provide Macaron with a GitHub token through the ``GITHUB_TOKEN``  environment variable.

    To obtain a GitHub Token:

    * Go to ``GitHub settings`` → ``Developer Settings`` (at the bottom of the left side pane) → ``Personal Access Tokens`` → ``Fine-grained personal access tokens`` → ``Generate new token``. Give your token a name and an expiry period.
    * Under ``"Repository access"``, choosing ``"Public Repositories (read-only)"`` should be good enough in most cases.

>     Now you should be good to run Macaron. For more details, see the documentation :ref:`here <prepare-github-token>`.

Reviewer 3: I don't think we should elaborate on this. If we are worried that the user does not know how to use Macaron, they should not read this page in the first place.

Others may have different ideas about keeping every tutorial to be standalone. I, however, don't think this tutorial should be a standalone one because it introduces an optional configuration rather than a concrete standalone use case. If the user doesn't understand the basic tutorial, I don't expect them to go to this page.

If we are afraid that they find this page by accident for whatever reason, I think it is probably best to say:

```
Prerequisite: You are expected to have gone through <link to the first tutorial>.
```

Otherwise, it just feels too noisy to me.

Reviewer 2: Agreed, this is a tutorial for a feature, not a use case, and so cannot be the user's first encounter with Macaron.

Author: Removed in https://github.com/oracle/macaron/pull/667/commits/7da9505dfcd2bd8b0b4e99832939a1701c75c7c8

------------------
Motivating example
------------------

Let's say you want to run the analysis on `the micronaut-projects/micronaut-core Github repository <https://github.com/micronaut-projects/micronaut-core>`_ at the tag ``4.3.10``.

Normally, this is how you would run Macaron:

.. code-block:: shell

  ./run_macaron.sh analyze -purl pkg:/github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps

> By default, Macaron will run all of its checks. However, there can be checks that are not applicable to ``micronaut-projects/micronaut-core``.

Reviewer 3: I think this sentence is a high-level description of the general cases.

```
By default, Macaron will run all of its checks. However, there can be checks that do not apply to the project of interest.
```

Because of this, I prefer it to be mentioned at the beginning, rather than after the example has been introduced.

Author: I have moved it to the beginning of the tutorial - https://github.com/oracle/macaron/pull/667/commits/44541c9e72aa223985fa36c7b0cfb3cdf77cd3f5

For example, the ``mcn_provenance_witness_level_one_1`` check (defined in :class:`ProvenanceWitnessL1Check <macaron.slsa_analyzer.checks.provenance_witness_l1_check.ProvenanceWitnessL1Check>`) is not applicable because ``micronaut-projects/micronaut-core`` doesn't generate and publish any `Witness provenance <https://github.com/testifysec/witness>`_. Therefore, we could exclude this check from running by performing the following steps.

1. Prepare a ``defaults.ini`` configuration file with the following content.

.. code-block:: ini

    [analysis.checks]
    exclude = mcn_provenance_witness_level_one_1
    include = *

In the ini configuration above:

* The ``exclude`` option contains the check ID ``mcn_provenance_witness_level_one_1``. The check with this ID will be excluded.
* The ``include`` option contains the `glob <https://docs.python.org/3/library/glob.html>`_ pattern ``*``. Checks whose ID matched this glob pattern will be included in the analysis. In this case, ``*`` matches all check IDs.

> 

Reviewer 3: We should state the eventual effect of the two options.

Author: By eventual effect, do you mean that the final effects on the run checks (e.g. `include = *` will include all checks currently available in Macaron) :thinking: ?

Reviewer 3: I think it should be mentioned that "With these two configuration options, all checks except for the excluded one (reference the name) will run."

Author: Added in https://github.com/oracle/macaron/pull/667/commits/b5781537fea25a57a2cffc1bf598e2ab5914baaa

2. Run Macaron analyze by providing the path to the custom ``defaults.ini`` configuration.

.. code-block:: shell

>   ./run_macaron.sh -dp <path_to_defaults_ini> analyze -purl pkg:/github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps

Reviewer 3: I think it's best to just decide on a fixed name and path for the config file and use it throughout the tutorial. I prefer users to just follow and not have to make additional decisions.

We may state explicitly in step 1 that

```
In your current directory, create a `defaults.ini` configuration file with the following content.
```
then use this `./defaults.ini` path throughout.

Author: Addressed in https://github.com/oracle/macaron/pull/667/commits/2d78fc7200f15506d4f8b80dbf681ec947db2f8e

>   ./run_macaron.sh -dp <path_to_defaults_ini> analyze -purl pkg:/github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps

Reviewer 3: The PURL here is malformed.
```suggestion
  ./run_macaron.sh -dp <path_to_defaults_ini> analyze -purl pkg:github.com/micronaut-projects/micronaut-core@v4.3.10 --skip-deps
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/3eed119b3ddbe737dc14d251ac104c6b7441e5cc

> In this run, the check ``mcn_provenance_witness_level_one_1`` mentioned above will not be run. After running the analyze command, we can view the data that Macaron has gathered about the ``micronaut-projects/micronaut-core`` repository at ``v4.3.10`` in an HTML report. Note that the result of the excluded check is not recorded in the Macaron HTML report (this also happens to JSON reports and the database).

Reviewer 3: Suggestion: Avoiding multiple nearby uses of the word `run` under different meanings may be best.

```suggestion
This time, the check ``mcn_provenance_witness_level_one_1`` runs. After the `analyze` command finishes, we can view the data that Macaron has gathered about the ``micronaut-projects/micronaut-core`` repository at ``v4.3.10`` in an HTML report. Note that the result of the excluded check is not recorded in the HTML reports, JSON reports, or the database.
```

Author: I believe the first sentence should be `This time, the check ``mcn_provenance_witness_level_one_1`` doesn't run.`

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/f72fc8ed825545d5a193083b711cd6a0720a0465

The two lists of ``exclude`` and ``include`` check IDs obtained from your configuration can be defined as:

* ``EXPLICIT_EXCLUDE`` = checks excluded from the user configuration.
* ``EXPLICIT_INCLUDE`` = checks included from the user configuration.

Checks in Macaron depend on each other. Therefore, when you want to exclude/include a check, you must be aware of how that will affect its transitive children/parents:

* If a check is excluded, all of the children which can be transitively reached from it will be excluded.
* If a check is included, all of the parents which can be transitively reached from it will be included.

We define the list of check IDs after extending to transitive parents/children as:

* ``EFFECTIVE_INCLUDED`` = ``EXPLICIT_INCLUDE`` set plus transitive parents
* ``EFFECTIVE_EXCLUDED`` = ``EXPLICIT_EXCLUDE`` set plus transitive children

The final list of checks that will run can be defined as:

* ``FINAL_INCLUDED`` = ``EFFECTIVE_INCLUDED`` minus ``EFFECTIVE_EXCLUDED``

-------------
More examples
-------------

^^^^^^^^^^^^^^^^^^^^^
Not running any check
^^^^^^^^^^^^^^^^^^^^^

> The following configurations will result in no check run and Macaron will return on error.

Reviewer 3: ```suggestion
The following configurations will result in no check running and Macaron will return on error.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/f72fc8ed825545d5a193083b711cd6a0720a0465

.. code-block:: ini

    [analysis.checks]
    exclude = *
    include = *

.. code-block:: ini

    [analysis.checks]
    exclude =
    include =

^^^^^^^^^^^^^^^^^^^^^^
Run a subset of checks
^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: ini

    [analysis.checks]
    exclude =
    include =
        mcn_provenance_witness_level_one_1
        mcn_trusted_builder_level_three_1

> This will results in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1`` and their transitive parents being run.

Reviewer 3: ```suggestion
This will result in ``mcn_provenance_witness_level_one_1``, ``mcn_trusted_builder_level_three_1`` and their transitive parents running.
```

Author: Fixed in https://github.com/oracle/macaron/pull/667/commits/f72fc8ed825545d5a193083b711cd6a0720a0465

> 

Reviewer 3: I think you may consider adding an example for "Disable a subset of checks" because this seems common enough.

Reviewer 3: It is also explicitly mentioned at the top of the page that the tutorial will show how to

> exclude checks that you don't want to run.

In that case, I think this is a justified suggestion.

Author: Added in https://github.com/oracle/macaron/pull/667/commits/d7273be7c8d8573605115450ec8e40f8455dd684
